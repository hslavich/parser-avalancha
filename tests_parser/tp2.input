fun esNat
  Zero -> True
  Suc(n) -> esNat(n)
  _ -> False

fun esLista
  Nil -> True
  Cons(_, xs) -> esLista(xs)
  _ -> False

fun zip : xs, ys -> _
  ? esLista(xs) and esLista(ys)
  Nil , _ -> Nil
  _ , Nil -> Nil
  Cons(x, xs) , Cons(y, ys) -> Cons(Par(x, y), zip(xs, ys))
