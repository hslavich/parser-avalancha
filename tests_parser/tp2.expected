[
  "program",
  [
    [
      "fun",
      "esNat",
      [
        "sig",
        [
          "_"
        ],
        "_"
      ],
      [
        "pre",
        [
          "true"
        ]
      ],
      [
        "post",
        [
          "true"
        ]
      ],
      [
        [
          "rule",
          [
            [
              "pcons",
              "Zero",
              []
            ]
          ],
          [
            "cons",
            "True",
            []
          ]
        ],
        [
          "rule",
          [
            [
              "pcons",
              "Suc",
              [
                [
                  "pvar",
                  "n"
                ]
              ]
            ]
          ],
          [
            "app",
            "esNat",
            [
              [
                "var",
                "n"
              ]
            ]
          ]
        ],
        [
          "rule",
          [
            [
              "pwild"
            ]
          ],
          [
            "cons",
            "False",
            []
          ]
        ]
      ]
    ],
    [
      "fun",
      "esLista",
      [
        "sig",
        [
          "_"
        ],
        "_"
      ],
      [
        "pre",
        [
          "true"
        ]
      ],
      [
        "post",
        [
          "true"
        ]
      ],
      [
        [
          "rule",
          [
            [
              "pcons",
              "Nil",
              []
            ]
          ],
          [
            "cons",
            "True",
            []
          ]
        ],
        [
          "rule",
          [
            [
              "pcons",
              "Cons",
              [
                [
                  "pwild"
                ],
                [
                  "pvar",
                  "xs"
                ]
              ]
            ]
          ],
          [
            "app",
            "esLista",
            [
              [
                "var",
                "xs"
              ]
            ]
          ]
        ],
        [
          "rule",
          [
            [
              "pwild"
            ]
          ],
          [
            "cons",
            "False",
            []
          ]
        ]
      ]
    ],
    [
      "fun",
      "zip",
      [
        "sig",
        [
          "xs",
          "ys"
        ],
        "_"
      ],
      [
        "pre",
        [
          "and",
          [
            "equal",
            [
              "app",
              "esLista",
              [
                [
                  "var",
                  "xs"
                ]
              ]
            ],
            [
              "cons",
              "True",
              []
            ]
          ],
          [
            "equal",
            [
              "app",
              "esLista",
              [
                [
                  "var",
                  "ys"
                ]
              ]
            ],
            [
              "cons",
              "True",
              []
            ]
          ]
        ]
      ],
      [
        "post",
        [
          "true"
        ]
      ],
      [
        [
          "rule",
          [
            [
              "pcons",
              "Nil",
              []
            ],
            [
              "pwild"
            ]
          ],
          [
            "cons",
            "Nil",
            []
          ]
        ],
        [
          "rule",
          [
            [
              "pwild"
            ],
            [
              "pcons",
              "Nil",
              []
            ]
          ],
          [
            "cons",
            "Nil",
            []
          ]
        ],
        [
          "rule",
          [
            [
              "pcons",
              "Cons",
              [
                [
                  "pvar",
                  "x"
                ],
                [
                  "pvar",
                  "xs"
                ]
              ]
            ],
            [
              "pcons",
              "Cons",
              [
                [
                  "pvar",
                  "y"
                ],
                [
                  "pvar",
                  "ys"
                ]
              ]
            ]
          ],
          [
            "cons",
            "Cons",
            [
              [
                "cons",
                "Par",
                [
                  [
                    "var",
                    "x"
                  ],
                  [
                    "var",
                    "y"
                  ]
                ]
              ],
              [
                "app",
                "zip",
                [
                  [
                    "var",
                    "xs"
                  ],
                  [
                    "var",
                    "ys"
                  ]
                ]
              ]
            ]
          ]
        ]
      ]
    ]
  ],
  []
]